BJS2-40651 Project Service : Vacancies - Update Vacancy Feature

drewbal opened the pull request:
Part 2/5 of Vacancy feature

drewbal committed:
Initial commit

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds VacancyDto

DTO: VacancyDto added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds Vacancy Mapper

DTO: VacancyDto optimized
Mapper: VacancyMapper added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds controller and create method

VacancyController: createVacancy method added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - create full vacancy from dto

VacancyService: toEntityFromDto method added (add fields ignored by mapper)
ProjectService: getProjectById metod added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - validate data to create the vacancy

VacancyValidator: checks the project is valid; role of the creator is valid;
TeamMemberService: returns the team member by id
DataValidationException: new Exception

drewbal committed:
BJS2-40651 Feature: Create Vacancy - creates new vacancy in system

VacancyService: creates a new vacancy;

drewbal committed:
BJS2-40651 Feature: Create Vacancy - sets status at creation

VacancyDto: VacationStatus field added;
VacancyService: sets OPEN status at creation;

drewbal committed:
BJS2-40651 Feature: Create Vacancy - review fixes

VacancyService: toEntityFromDto is private;
VacancyController: path fixed;
Other: blank lines removed

drewbal committed:
BJS2-40651 Feature: Update Vacancy - controller

VacancyController: new method to update the existing vacancy status

drewbal committed:
BJS2-40651 Feature: Update Vacancy - dto update

VacancyDto: upgradeBy field added
Other: VacancyMapper class moved to other package

drewbal committed:
BJS2-40651 Feature: Update Vacancy - validator update

VacancyValidator: checks the role od updater and the number of candidates for vacancy closure
Exception: InsufficientCandidatesException added

drewbal committed:
BJS2-40651 Feature: Update Vacancy - can update vacancy status to CLOSED

VacancyService: updateVacancyStatus method that changes the status of the vacancy based on candidates amount criteria

drewbal committed:
Merge branch 'Dragon-master-stream7' into BJS2-40651-update-vacancy

YTXIRE commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("api/vacancies")
```
Те же грабли)
Не забывай про /api

И давай сразу версионирование сделаем v1

Т.е. /api/v1/vacancies

YTXIRE commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,58 @@
+        return getVacancyById(vacancyId).getCandidates();
+    }
+
+    public Vacancy getVacancyById(long vacancyId) {
```
Сделай так же private

drewbal replied:
Мне кажется, что лучше его оставить публичным. Он не сугубо утилитарный в рамках сервиса. Если где-то в проекте понадобится поднять вакансию из бд, то будет использоваться именно этот метод. Поэтому считаю, что такие базовые методы должны быть публичными в каждом сервисе

YTXIRE commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,58 @@
+
+    public Vacancy getVacancyById(long vacancyId) {
+        return vacancyRepository.findById(vacancyId).orElseThrow(
+                () -> new EntityNotFoundException(String.format("Vacancy not found by id: %s", vacancyId)));
```
Я как раз добавил обработчик ошибок

drewbal replied:
Эх, надо было его тебе сделать отдельным коммитом и  в мастер сразу отправить

drewbal committed:
BJS2-40651 Feature: Update Vacancy - review fixes

VacancyController: api v1 added to path
VacancyService: @Transactional annotation added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - review fixes

VacancyValidator: logic changed, VacancyRepository injected
VacancyService: validation fix
ProjectValidator: basic validation added

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/main/java/faang/school/projectservice/validator/VacancyValidator.java
#	src/test/java/faang/school/projectservice/validator/VacancyValidatorTest.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy - review fixes

Other: minor conflict fixes

drewbal committed:
Merge branch 'Dragon-master-stream7' into BJS2-40651-create-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/exception/DataValidationException.java
#	src/main/java/faang/school/projectservice/service/TeamMemberService.java

drewbal committed:
BJS2-40651 Feature: Create Vacancy - merge and conflicts solving

Minor fixes

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/mapper/VacancyMapper.java
#	src/main/java/faang/school/projectservice/validator/VacancyValidator.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/validator/VacancyValidatorTest.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy - merge and conflicts solving

Minor fixes

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

VacancyController: logging added
ServiceController: logging added
GlobalExceptionHandler: fixed wrong import, logging added
Other: spaces fixed

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/service/VacancyService.java

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

VacancyController: logging added
ServiceController: logging added
GlobalExceptionHandler: InsufficientCandidates exception added
Other: Exception imports fixed

drewbal committed:
BJS2-40651 Feature: Create Vacancy: optimization

NewVacancyDto: new DTO for vacancy creation
VacancyController: createVacancy method accepts new dto with minimum required fields
VacancyMapper: takes new vacancy DTO, returns default DTO
VacancyService: create method optimized for new dto
VacancyValidator: role validation methods united

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/VacancyController.java
#	src/main/java/faang/school/projectservice/mapper/VacancyMapper.java
#	src/main/java/faang/school/projectservice/service/VacancyService.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy: optimization

VacancyUpdateDto: new DTO for vacancy update
VacancyController: updateVacancy method accepts new dto with minimum required fields
VacancyService: update method optimized for new dto
VacancyValidator: obsolete role validation method removed

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

Vacancy: createdBy -> createdById, updatedBy ->updatedById
NewVacancyDto: validation annotations fixed
VacancyDto: renamed in VacancyResponseDto to improve logic
Other: exception messages fixed

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/VacancyController.java
#	src/main/java/faang/school/projectservice/dto/vacancy/VacancyDto.java
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java
#	src/test/java/faang/school/projectservice/validator/VacancyValidatorTest.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy: review fixes

Fixes from merge of Create Vacancy Feature
